### PR TITLE
assorted subtyping tests

### PR DESCRIPTION
**Pull request by felixrehren**
c.f. master list #19998
tests for closed issues #2552 #8625 #8915 #11407
and #12814 #16922 #17943 #18892 (#19159 #19413) #18985 #19041

[Edit: misunderstood method overwriting as in #18892, thanks yuyichao for correction, now has a test]

### CODE REVIEW

**Comment by tkelman** [comment]
these should use the new syntax

**Comment by yuyichao** [comment]
Why should it error? You defined two methods that overrides each other and both of them matches the signature you are calling.

**Comment by felixrehren** [comment]
Like what? `type D8915{<:Union{Float32,Float64}}` instead of `type D8915{T <: Union{Float32,Float64}}` ?

**Comment by felixrehren** [comment]
I don't understand the error message
```
  LoadError: syntax: 
  ERROR: deprecated syntax "inner constructor D8915(...) around /tmp/julia/share/julia/test/subtype.jl:932".
  Use "D8915{T}(...) where T" instead.
```
and can't work out from `NEWS.md` what the new syntax should be :/

File: test/subtype.jl [hunk 1]
```diff
@@ -939,3 +939,36 @@ f11407{K,V}(::Dict{K,V},::Dict{Any,V}) = 1
...
+@test f18892(0) == 2
+
+# #19041
+mutable struct A19041{K} end
+A19041(K::Type) = A19041{K}()
+struct B19041{T} end
+f19041{K}(x::A19041{K}) = deepcopy(x)
+@test f19041(A19041(B19041)) == A19041{B19041}()
+
+# #19985
```
Comment (tkelman) [inline test/subtype.jl]: typo?

**Comment by yuyichao** [comment]
> and can't work out from NEWS.md what the new syntax should be :/

Have you tried doing exactly what the deprecation warning says?

**Comment by StefanKarpinski** [comment]
> Have you tried doing exactly what the deprecation warning says?

It's on the second line of the error message in case you missed it (I did on first reading).

File: test/subtype.jl [hunk 2]
```diff
@@ -915,3 +915,59 @@ ftwoparams(::TwoParams{<:Real,<:Real}) = 3
...
+f11407{K,V}(::Dict{K,V},::Dict{K,Any}) = 2
+d11407 = Dict{Any,Any}()
+@test_throws MethodError f11407(d11407,d11407)
+
+# #12814
+abstract type A12814{N, T} end
+struct B12814{N, T} <: A12814{N, T}
+    x::NTuple{N, T}
+end
+Base.call{T <: A12814, X <: Array}(::Type{T}, x::X) = 1
```
Comment (tkelman) [inline test/subtype.jl]: call overloading in this style is deprecated

Comment (felixrehren) [inline test/subtype.jl]: I saw that when I tried this is in 0.5, but couldn't figure out the transformation (and the warning had disappeared in 0.6?). How should this be spelled now?

Comment (yuyichao) [inline test/subtype.jl]: You should follow the deprecation warning on 0.5.

`(::Type{T}){T<:A12814,X<:Array}(x::X) = 1`

**Comment by felixrehren** [comment]
Thanks all for the help!
@yuyichao I tried to follow the depreciation warning, but I don't understand it (what inner constructor?), and my attempts to fix it haven't worked. I would be grateful for any further pointers / a fix!

**Comment by yuyichao** [comment]
> (what inner constructor?)

The meaning of `inner cconstructor` should be very unambiguous: http://docs.julialang.org/en/latest/manual/constructors.html#Inner-Constructor-Methods-1

> and my attempts to fix it haven't worked.

What didn't work? The depwarn shouldn't be printed anymore?

**Comment by felixrehren** [comment]
Writing `D8915{T}(a) where T = 1` instead of `D8915(a) = 1` didn't work: the depwarn is gone, but replaced by
```LoadError: syntax: "1" is not a valid function argument name
  while loading /tmp/julia/share/julia/test/subtype.jl, in expression starting on line 935
```
and
```
Got an exception of type LoadError outside of a @test
ERROR: LoadError: type UnionAll has no field parameters
```
on two of the Travis builds.

**Comment by yuyichao** [comment]
Please check the line number. That's not the same line at all.

https://github.com/JuliaLang/julia/pull/20627/files#diff-4cf37a18308090507d83e9d96dd01692R935 is wrong.

**Comment by tkelman** [comment]
I don't think that deprecation message is all that obvious, fwiw

**Comment by yuyichao** [comment]
What would be a more obvious one?

**Comment by tkelman** [comment]
printing the actual contents of the line/signature that caused the warning instead of a `(...)` abbreviation of it

**Comment by yuyichao** [comment]
The abbreviated part is not important since it does not need to be changed at all. It should be ok to print it if it's short but when it's longer it'll make it much harder to spot the actual change that needs to be made.

**Comment by tkelman** [comment]
it is important if it makes it easier to identify what actually needs changing. right now it's not very clear, like when there are multiple inner constructors

**Comment by yuyichao** [comment]
What about multiple inner constructors? If none of them have type parameters they need to be changed in exactly the same way and are printed as such in the deprecation warning?

**Comment by tkelman** [comment]
then you should get the warning once per constructor, otherwise fixing one will only change the line number of the warning (if you're lucky) so you don't get much feedback that you partially fixed it

**Comment by yuyichao** [comment]
> then you should get the warning once per constructor

You do?

```julia
julia> struct A{T}
           A() = 1
           A(a) = 2
       end

WARNING: deprecated syntax "inner constructor A(...) around REPL[1]:2".
Use "A{T}(...) where T" instead.

WARNING: deprecated syntax "inner constructor A(...) around REPL[1]:3".
Use "A{T}(...) where T" instead.
```

**Comment by yuyichao** [comment]
Just noticed that the inner constructor with type parameters is not printed correctly.

```julia
julia> struct C{T}
           C{T2}(a::T2) = T2
       end

WARNING: deprecated syntax "inner constructor C(...) around REPL[3]:2".
Use "C{T}(...) where T" instead.
```

Should say

```julia
WARNING: deprecated syntax "inner constructor C{T2}(...) around REPL[3]:2".
Use "C{T}(...) where {T,T2}" instead.
```

Though this gets a lot trickier for

```julia
struct D{T}
    D{T}(::T) = 1
end
```

which should probably suggest sth like `D{_}(::T) where {_,T} = T`.

**Comment by felixrehren** [comment]
What should the correct version of lines 930 - 935 be?

**Comment by yuyichao** [comment]
```julia
mutable struct D8915{T <: Union{Float32,Float64}}
    D8915{T}(a) where T = 1
    D8915{T}(a::Int) where T = 2
end
@test D8915{Float64}(1) == 2
```

**Comment by tkelman** [comment]
preferably with curly braces `where {T}` for a bit nicer readability

**Comment by felixrehren** [comment]
Is there a problem with remote exceptions? Not sure what to make of this message in the logs, seems to be a different remote worker that failed in each case
```
Error in testset subtype:
Error During Test
  Test threw an exception of type RemoteException
  Expression: subtype
  RemoteException(1, CapturedException(CompositeException(Any[CapturedException(ErrorException("Error deserializing a remote exception from worker 8\nRemote(original) exception of type Base.Test.TestSetException\nRemote stacktrace : "), Any[(fini...
```

**Comment by yuyichao** [comment]
Yes there are some issues with test result serialization so it doesn't actually show the error.

OTOH, please run the tests locally first before pushing.

**Comment by pabloferz** [comment]
Rebased (hope you don't mind)

**Comment by tkelman** [comment]
likely needs "fixing whatever was broken about the tests" too in addition to just a rebase, so this'll probably fail at the moment

**Comment by felixrehren** [comment]
Thanks for rebase @pabloferz! ❤️ Fixed tests, builds build. LGTM?

**Comment by Keno** [comment]
Bump @JeffBezanson

**Comment by Keno** [comment]
Bump @JeffBezanson are any of these still relevant?

**Pull request merged**